core/example/longrange/ewald.cpp

```diff
@@ -0,0 +1,270 @@
+/****************************************************************************
+ * Copyright (c) 2018-2019 by the Cabana authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the Cabana library. Cabana is distributed under a   *
+ * BSD 3-clause license. For the licensing terms see the LICENSE file in    *
+ * the top-level directory.                                                 *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include "definitions.h"
+#include "ewald.h"
+#include <cmath>
+
+TEwald::TEwald(double accuracy, ParticleList particles, double lx, double ly, double lz)
+{
+  _r_max = 0.0;
+  tune(accuracy,particles,lx, ly, lz);
+}
+
+TEwald::TEwald(double alpha, double r_max, double k_max)
+{
+  _alpha = alpha;
+  _r_max = r_max;
+  _k_max = k_max;
+  _k_max_int[0] = std::ceil(_k_max);
+  _k_max_int[1] = std::ceil(_k_max);
+  _k_max_int[2] = std::ceil(_k_max);
+}
+
+void TEwald::tune(double accuracy, ParticleList particles, double lx, double ly, double lz)
+{
+  typedef Kokkos::MinLoc<double, int> reducer_type;
+  typedef reducer_type::value_type value_type;
+  value_type error_estimate;
+
+  auto q = particles.slice<Charge>();
+  double q_sum;  
+
+  const int N_alpha = 200;
+  const int N_k = 2000;
+
+  const int n_max = particles.size();
+
+  // calculate sum of charge squares
+  Kokkos::parallel_reduce( n_max, KOKKOS_LAMBDA( const int idx, double& q_part )
```

Comment:
Note that if you ever decide to use Cuda UVM memory you will need to fence after parallel kernels.

Reply:
To address this, would you recommend adding a `Kokkos::fence()` after parallel kernels for all cases, or just in the case where `MemorySpace` is `Cabana::CudaUVMSpace`?